guides/inventing-service.md

```diff
@@ -0,0 +1,919 @@
+since they wrap another `Service` and transform the request or response in some
+way. Whereas types like `RequestHandler` is typically called a "leaf service"
+since it sits at the leaves of a tree of nested services. The actual responses
+are normally produced in leaf services and modified by middlewares.
+
+The only thing left to talk about is "backpressure" and `poll_ready`.
+
+## Backpressure
+
+Imagine you wanted to write a rate limit middleware that wraps a `Service` and
+puts a limit on the maximum number of concurrent requests the underlying service
+will receive. This would be useful if you had some service that had a hard upper
+limit on the amount of load it could handle.
+
+With our current `Service` trait we don't really have a good way to implement
+something like that. We could try:
+
+```rust
+impl<R, T> Service<R> for ConcurrencyLimit<T> {
+    fn call(&mut self, request: R) -> Self::Future {
+        // 1. Check a counter for the number of requests currently being
+        //    processed.
+        // 2. If there is capacity left send the request to `T`
+        //    and increment the counter.
+        // 3. If not somehow wait until capacity becomes available.
+        // 4. When the response has been produced, decrement the counter.
+    }
+}
+```
+
+If there is no capacity left we have to wait and somehow get notified when
+capacity becomes available. Additionally we have to keep the request in memory
+while we're waiting (also called buffering). It would be more robust to only
+allocate space for the request when we are sure the service has capacity to
+handle it. Otherwise we risk using a lot of memory buffering requests while we
+wait for our service to become ready.
+
+In a way it would be nice if `Service` had a method like this:
+
+```rust
+trait Service<R> {
+    async fn ready(&mut self);
+}
+```
+
+`ready` would be an async function that completes when the service has capacity
+enough to receive one new request. We would then require users to first call
+`service.ready().await` before doing `service.call(request).await`.
+
+Separating "calling the service" from "reserving capacity" also unlocks new use
+cases such as being able to maintain a set of "ready services" that we keep up
+to date in the background such that when a request arrives we already have a
+ready service to send it to and don't have to first wait for it to become ready.
+
+With this setup `ConcurrencyLimit` could track capacity inside `ready` and not
+allow users to call `call` until there is sufficient capacity.
+
+`Service`s that don't care about capacity can just return immediately from
+`ready` or delegate to some inner `Service` they're wrapping.
+
+However we still cannot define async functions inside traits. We could add
+another associated type to `Service` called `ReadyFuture` but having to return a
+`Future` will give us the same lifetime issues we've run into before. It would
+be nice if there was someway around that.
```

Comment:
```suggestion
However, we still cannot define async functions inside traits. We could add
another associated type to `Service` called `ReadyFuture`, but having to return a
`Future` will give us the same lifetime issues we've run into before. It would
be nice if there was some way around that.
```